selfservice/flow/login/flow.go

```diff
@@ -231,6 +231,10 @@ func (f *Flow) EnsureInternalContext() {
 	}
 }
 
+func (f *Flow) GetInternalContext() *sqlxx.JSONRawMessage {
+	return &f.InternalContext
```

Comment:
Why return a pointer?

Reply:
So that we can mutate the internal context as well.

Reply:
OK I see - you're using it here:

```
	*flow.GetInternalContext() = bytes
```

I guess that's fine (it works), just a bit unusual (would expect `SetInternalContext()`)

Reply:
So what I mean by that is that it's unusual to have pointers to slices. Here's why:

```go
package main

func main() {
	var a []byte
	var b *[]byte
	b = &a
	if b == nil {
		print("nil")
	}
	print((*b)[0]) // panics
}
```

This problem exists here where you check whether the pointer is set, but not whether the internal context is actually set:

```
func SetDuplicateCredentials(flow InternalContexter, creds DuplicateCredentialsData) error {
	if flow.GetInternalContext() == nil {
		flow.EnsureInternalContext()
	}
	bytes, err := sjson.SetBytes(
		*flow.GetInternalContext(),
		internalContextDuplicateCredentialsPath,
		creds,
	)
	if err != nil {
		return err
	}
	*flow.GetInternalContext() = bytes

	return nil
}
```

Since `internalContext` is a []byte per default (and thus nil per default), you'll probably run into some errors here  in `SetBytes`.

Reply:
I am calling `flow.EnsureInternalContext()` if the slice is nil to make sure it is always allocated.

In your code example, the panic is not because the pointer points to nothing (it points to `&a`, which is on the stack), but because it is an empty slice and accessing the first element panics.

---

I can still add a SetInternalContext if you want, just to make the code easier to read. WDYT?

Reply:
fixed

Reply:
Yes that's true, what I meant is this:

```
	if flow.GetInternalContext() == nil { // might evaluate to trfalseue because the pointer is not nil but the memory address is nil
		flow.EnsureInternalContext() // would set context to string "{}" but won't because pointer is not-nil even though value is nil
	}
	
	bytes, err := sjson.SetBytes(
		*flow.GetInternalContext(),
		internalContextDuplicateCredentialsPath,
		creds,
	)
	// err is now non-nil because `*flow.GetInternalContext()` is nil which `SetBytes` can't deal with a nil byte slice
```

However, I realized that `SetBytes()` can deal with a nil slice and won't return an error. So this was indeed fine, at the same time I think that the setter is a clearer pattern as it doesn't have this "hidden" nil-pointer versus nil-slice-value meaning.